Add internal model for plugin management

Jack251970 opened the pull request:
# Add internal model for plugin management

Let us do not rely `Plugin Manager` plugin for plugin management in Flow Launcher. Flow Launcher will fall back into a bad state if  this plugin is deleted.

Resolve #3555, #3236.

Jack251970 commented on Flow.Launcher/ViewModel/PluginStoreItemViewModel.cs:
```diff
@@ -59,11 +69,223 @@ public string Category
-        private void ShowCommandQuery(string action)
+        private async Task ShowCommandQueryAsync(string action)
+        {
+            switch (action)
```
We need to ignore default

Jack251970 commented on Flow.Launcher/ViewModel/PluginStoreItemViewModel.cs:
```diff
@@ -59,11 +69,223 @@
+                // check if user cancelled download before installing plugin
+                if (cts.IsCancellationRequested)
+                {
+                    return;
+                }
+                else
+                {
+                    if (!File.Exists(filePath))
+                    {
+                        throw new FileNotFoundException($"Plugin {newPlugin.ID} zip file not found at {filePath}", filePath);
+                    }
+
+                    App.API.InstallPlugin(newPlugin, filePath);
+
+                    if (!newPlugin.IsFromLocalInstallPath)
+                    {
+                        File.Delete(filePath);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                App.API.LogException(ClassName, "Failed to install plugin", e);
+                App.API.ShowMsgError(App.API.GetTranslation("ErrorInstallingPlugin"));
+            }
+
+            if (Settings.AutoRestartAfterChanging)
+            {
+                App.API.RestartApp();
+            }
+            else
+            {
+                App.API.ShowMsg(
+                    App.API.GetTranslation("installbtn"),
+                    string.Format(
+                        App.API.GetTranslation(
+                            "InstallSuccessNoRestart"),
+                        newPlugin.Name));
+            }
+        }
```
Good!

Jack251970 commented on Flow.Launcher/ViewModel/PluginStoreItemViewModel.cs:
```diff
@@ -59,11 +69,228 @@ public string Category
+                case "update":
+                    await UpdatePluginAsync(_newPlugin, _oldPluginPair.Metadata);
+                    break;
+                default:
```
Just ignore that

Jack251970 commented on Flow.Launcher.Core/Plugin/PluginManager.cs:
```diff
@@ -704,6 +915,59 @@ internal static async Task UninstallPluginAsync(PluginMetadata plugin, bool remo
+            if (pieces.Length < 4)
+                return false;
+
+            var author = pieces[3];
```
no need

jjw24 commented on Flow.Launcher.Core/Plugin/PluginManager.cs:
```diff
@@ -557,6 +561,213 @@ public static async Task UninstallPluginAsync(PluginMetadata plugin, bool remove
             await UninstallPluginAsync(plugin, removePluginFromSettings, removePluginSettings, true);
         }
 
+        public static async Task InstallPluginAndCheckRestartAsync(UserPlugin newPlugin)
```
I think we need to put these code in a separate file- PluginSetupManager. This is too much code to be in one file.

jjw24 replied:
It's not helper though, it's the core of how installation(uninstallation) is done, leave it in core/plugin and call it `PluginInstaller`.

Jack251970 replied:
Changed as required

Jack251970 commented on Flow.Launcher/Helper/PluginInstallationHelper.cs:
```diff
@@ -0,0 +1,283 @@
+                {
+                    File.Delete(filePath);
+                }
+            }
```
Good!

Jack251970 commented on Flow.Launcher/SettingPages/ViewModels/SettingsPanePluginStoreViewModel.cs:
```diff
@@ -1,7 +1,10 @@
+using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
+using System.Windows.Forms;
```
Good!

Jack251970 commented on Flow.Launcher/Helper/PluginInstallationHelper.cs:
```diff
@@ -0,0 +1,283 @@
+            var pluginJsonPath = archive.Entries.FirstOrDefault(x => x.Name == "plugin.json") ??
+                throw new FileNotFoundException("The zip file does not contain a plugin.json file.");
+            var pluginJsonEntry = archive.GetEntry(pluginJsonPath.ToString()) ??
+                throw new FileNotFoundException("The zip file does not contain a plugin.json file.");
```
Good!

Jack251970 commented on Flow.Launcher/Helper/PluginInstallationHelper.cs:
```diff
@@ -0,0 +1,277 @@
+                throw new FileNotFoundException("The zip file does not contain a plugin.json file.");
+
+            using Stream stream = pluginJsonEntry.Open();
+            plugin = JsonSerializer.Deserialize<UserPlugin>(stream);
```
No need to do so. It is already inside a try-catch sentence

Jack251970 commented on Flow.Launcher/Helper/PluginInstallationHelper.cs:
```diff
@@ -0,0 +1,283 @@
+        {
+            var exceptionHappened = false;
+            await App.API.ShowProgressBoxAsync(prgBoxTitle,
+                async (reportProgress) =>
+                {
+                    if (reportProgress == null)
+                    {
+                        // when reportProgress is null, it means there is exception with the progress box
+                        // so we record it with exceptionHappened and return so that progress box will close instantly
+                        exceptionHappened = true;
+                        return;
+                    }
+                    else
+                    {
+                        await App.API.HttpDownloadAsync(downloadUrl, filePath, reportProgress, cts.Token).ConfigureAwait(false);
+                    }
+                }, cts.Cancel);
```
No need

Jack251970 commented on Flow.Launcher.Core/Plugin/PluginInstaller.cs:
```diff
@@ -0,0 +1,287 @@
+        UserPlugin plugin;
+        try
+        {
+            using ZipArchive archive = ZipFile.OpenRead(filePath);
+            var pluginJsonEntry = archive.Entries.FirstOrDefault(x => x.Name == "plugin.json") ??
+                throw new FileNotFoundException("The zip file does not contain a plugin.json file.");
+
+            using Stream stream = pluginJsonEntry.Open();
+            plugin = JsonSerializer.Deserialize<UserPlugin>(stream);
```
no need since it is inside a try-catch sentence

jjw24 reviewed (changes requested):
Just found a couple minor issues:
1. After uninstalling pm, store no longer fetches plugin until clicking on refresh.
2. Using pm install/uninstall does not automatically restart flow even though the option is set
3. Store install/uninstall same plugin without restart shows error message error, should say already installed/uninstalled
4. pm install/uninstall same plugin without restart says correct message but another message also pops up to say it's successfully installed
5. pm uninstall same plugin without restart says correct message but string interpolation is not working

Jack251970 commented:
@jjw24 
> 1. After uninstalling pm, store no longer fetches plugin until clicking on refresh.

Fixed. This is because originally, Flow rely on pm to initialize the manifest. So we need to explicitly call update when app is loading.

> 2. Using pm install/uninstall does not automatically restart flow even though the option is set

You need to change the option in pm setting panel instead of Flow general setting page.

> 3. Store install/uninstall same plugin without restart shows error message error, should say already installed/uninstalled

Fixed. Now Flow will show msg when encountering those issues.

> 4. pm install/uninstall same plugin without restart says correct message but another message also pops up to say it's successfully installed

Fixed. Now Flow will show msg when encountering those issues.

> 5. pm uninstall same plugin without restart says correct message but string interpolation is not working

I cannot reproduce that. Please give the specific message.

Jack251970 commented on Flow.Launcher.Core/Plugin/PluginInstaller.cs:
```diff
@@ -0,0 +1,317 @@
+            await API.ShowProgressBoxAsync(prgBoxTitle,
+                async (reportProgress) =>
+                {
+                    if (reportProgress == null)
+                    {
+                        // when reportProgress is null, it means there is exception with the progress box
+                        // so we record it with exceptionHappened and return so that progress box will close instantly
+                        exceptionHappened = true;
+                        return;
+                    }
+                    else
+                    {
+                        await API.HttpDownloadAsync(downloadUrl, filePath, reportProgress, cts.Token).ConfigureAwait(false);
+                    }
```
No need to do

jjw24 commented:
4. This is not fixed- pm install/uninstall same plugin without restart says correct message but another message also pops up to say it's successfully installed. It should only one pop up notification to say that the plugin already installed/uninstalled.
6. pm uninstall plugin with option set to no restart, the plugin remains usable after the uninstall (can be triggered with action keyword)
7. Change the notification message title about plugin already installed/uninstalled from 'Fail to install <plugin name>' to '<plugin name> is already installed'. Do the same for uninstalling notification title.
8. Why are there two options for managing restarting after install/uninstall?
9 pm uninstall plugin twice with option set to no-restart also gives two notification, the correct one and the other is that the plugin is uninstalled successfully (and this shouldn't be shown because already uninstalled. Align behaviour with how Store uninstall pl

jjw24 commented:
```
4. This is not fixed- pm install/uninstall same plugin without restart says correct message but another message also pops up to say it's successfully installed. It should only one pop up notification to say that the plugin already installed/uninstalled.

6. pm uninstall plugin with option set to no restart, the plugin remains usable after the uninstall (can be triggered with action keyword)

7. Change the notification message title about plugin already installed/uninstalled from 'Fail to install <plugin name>' to '<plugin name> is already installed'. Do the same for uninstalling notification title.

8. Why are there two options for managing restarting after install/uninstall?

9. pm uninstall plugin twice with option set to no-restart also gives two notification, the correct one and the other is that the plugin is uninstalled successfully (and this shouldn't be shown because already uninstalled. Align behaviour with how Store uninstall plugin twice)
```

jjw24 commented:
8. Can we then change their title description to distinguish Store ve PM please. Also PM should also default to no restart iirc.

jjw24 commented:
10. Store uninstall plugin with option set to no restart, when using the plugin's action keyword, query window seems unresponsive, e.g. remove bookmark plugin and use `e google`
11. pm uninstall plugin with option set to no restart, when using the plugin's action keyword, query window seems unresponsive, e.g. remove bookmark plugin and use `e google`

Jack251970 commented:
> 10. Store uninstall plugin with option set to no restart, when using the plugin's action keyword, query window seems unresponsive, e.g. remove bookmark plugin and use `e google`
> 11. pm uninstall plugin with option set to no restart, when using the plugin's action keyword, query window seems unresponsive, e.g. remove bookmark plugin and use `e google`

Sorry, I cannot reproduce that. Could you please share your setting file?

jjw24 commented:
>Sorry, I cannot reproduce that. Could you please share your setting file and detailed steps?

I just did a fresh install, uninstalled bookmarks then used `b` like `b <my bookmark>`, looks like it's still tried to trigger with the action keyword.

![pm_uninstall_issue](https://github.com/user-attachments/assets/6c0c86d8-969d-4323-bde8-cd5609675873)


>Since you are using no restart option, I think your issue is possibly related to modified plugins. I wonder if https://github.com/Flow-Launcher/Flow.Launcher/pull/3795 can help since it improves modified plugin management.

Same issue exists there.

Could you try reproducing again please.

Jack251970 commented:
> > Sorry, I cannot reproduce that. Could you please share your setting file and detailed steps?
> 
> I just did a fresh install, uninstalled bookmarks then used `b` like `b <my bookmark>`, looks like it's still tried to trigger with the action keyword.
> 
> ![pm_uninstall_issue](https://private-user-images.githubusercontent.com/26427004/462882983-6c0c86d8-969d-4323-bde8-cd5609675873.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE3OTIyNjQsIm5iZiI6MTc1MTc5MTk2NCwicGF0aCI6Ii8yNjQyNzAwNC80NjI4ODI5ODMtNmMwYzg2ZDgtOTY5ZC00MzIzLWJkZTgtY2Q1NjA5Njc1ODczLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA2VDA4NTI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzNzM4ZTVlNTYxNTQ4M2E0ZWEyODEwZDI2NTE3ZDI1ODgxYTY4N2I0N2ZiYTY1MDE1NGMzYWUxNGJjMTg0MTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.b-jjWBInEC2FqknulywQ4Z3-Rt-OQcHkrNmhGPwsvGs) [ ![pm_uninstall_issue](https://private-user-images.githubusercontent.com/26427004/462882983-6c0c86d8-969d-4323-bde8-cd5609675873.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE3OTIyNjQsIm5iZiI6MTc1MTc5MTk2NCwicGF0aCI6Ii8yNjQyNzAwNC80NjI4ODI5ODMtNmMwYzg2ZDgtOTY5ZC00MzIzLWJkZTgtY2Q1NjA5Njc1ODczLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA2VDA4NTI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzNzM4ZTVlNTYxNTQ4M2E0ZWEyODEwZDI2NTE3ZDI1ODgxYTY4N2I0N2ZiYTY1MDE1NGMzYWUxNGJjMTg0MTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.b-jjWBInEC2FqknulywQ4Z3-Rt-OQcHkrNmhGPwsvGs) ](https://private-user-images.githubusercontent.com/26427004/462882983-6c0c86d8-969d-4323-bde8-cd5609675873.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE3OTIyNjQsIm5iZiI6MTc1MTc5MTk2NCwicGF0aCI6Ii8yNjQyNzAwNC80NjI4ODI5ODMtNmMwYzg2ZDgtOTY5ZC00MzIzLWJkZTgtY2Q1NjA5Njc1ODczLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA2VDA4NTI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzNzM4ZTVlNTYxNTQ4M2E0ZWEyODEwZDI2NTE3ZDI1ODgxYTY4N2I0N2ZiYTY1MDE1NGMzYWUxNGJjMTg0MTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.b-jjWBInEC2FqknulywQ4Z3-Rt-OQcHkrNmhGPwsvGs) [ ](https://private-user-images.githubusercontent.com/26427004/462882983-6c0c86d8-969d-4323-bde8-cd5609675873.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE3OTIyNjQsIm5iZiI6MTc1MTc5MTk2NCwicGF0aCI6Ii8yNjQyNzAwNC80NjI4ODI5ODMtNmMwYzg2ZDgtOTY5ZC00MzIzLWJkZTgtY2Q1NjA5Njc1ODczLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA2VDA4NTI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzNzM4ZTVlNTYxNTQ4M2E0ZWEyODEwZDI2NTE3ZDI1ODgxYTY4N2I0N2ZiYTY1MDE1NGMzYWUxNGJjMTg0MTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.b-jjWBInEC2FqknulywQ4Z3-Rt-OQcHkrNmhGPwsvGs)
> 
> > Since you are using no restart option, I think your issue is possibly related to modified plugins. I wonder if #3795 can help since it improves modified plugin management.
> 
> Same issue exists there.
> 
> Could you try reproducing again please.

Fixed in #3802. Please check.

jjw24 commented:
@Jack251970 please resolve conflict from merging of PR 3795

Jack251970 commented on Flow.Launcher.Core/Plugin/PluginManager.cs:
```diff
@@ -561,33 +561,46 @@ public static bool PluginModified(string id)
+        public static async Task<bool> UpdatePluginAsync(PluginMetadata existingVersion, UserPlugin newVersion, string zipFilePath)
         {
-            InstallPlugin(newVersion, zipFilePath, checkModified: false);
-            await UninstallPluginAsync(existingVersion, removePluginFromSettings: false, removePluginSettings: false, checkModified: false);
-            ModifiedPlugins.Add(existingVersion.ID);
+            if (PluginModified(existingVersion.ID))
+            {
+                API.ShowMsgError(string.Format(API.GetTranslation("pluginModifiedAlreadyTitle"), existingVersion.Name),
+                    API.GetTranslation("pluginModifiedAlreadyMessage"));
```
No need since Update function we cannot call this check inside

Jack251970 commented:
> @Jack251970 please resolve conflict from merging of PR 3795

Done!
